stackslib/src/chainstate/nakamoto/coordinator/mod.rs

```diff
@@ -0,0 +1,844 @@
+        let mut sortitions_to_process = VecDeque::new();
+        let last_processed_ancestor = loop {
+            if let Some(found_sortition) = self.sortition_db.is_sortition_processed(&cursor)? {
+                debug!(
+                    "Ancestor sortition {} of block {} is processed",
+                    &found_sortition, &cursor
+                );
+                break found_sortition;
+            }
+
+            let current_block =
+                BurnchainDB::get_burnchain_block(&self.burnchain_blocks_db.conn(), &cursor)
+                    .map_err(|e| {
+                        warn!(
+                            "ChainsCoordinator: could not retrieve  block burnhash={}",
+                            &cursor
+                        );
+                        Error::NonContiguousBurnchainBlock(e)
+                    })?;
+
+            debug!(
+                "Unprocessed block: ({}, {})",
+                &current_block.header.block_hash.to_string(),
+                current_block.header.block_height
+            );
+
+            let parent = current_block.header.parent_block_hash.clone();
+            sortitions_to_process.push_front(current_block);
+            cursor = parent;
+        };
+        Ok((last_processed_ancestor, sortitions_to_process))
+    }
+
+    /// Process the next-available burnchain block, if possible.
+    /// Burnchain blocks can only be processed for the last-known PoX reward set, which is to say,
+    /// burnchain block processing can be blocked on the unavailability of the next PoX anchor
+    /// block.  If the next PoX anchor block is not available, then no burnchain block processing
+    /// happens, and this function returns false.  It returns true otherwise.
+    ///
+    /// Returns Err(..) if an error occurred while processing (i.e. a DB error).
+    pub fn handle_new_nakamoto_burnchain_block(&mut self) -> Result<bool, Error> {
+        // highest burnchain block we've downloaded
+        let canonical_burnchain_tip = self.burnchain_blocks_db.get_canonical_chain_tip()?;
+
+        debug!("Handle new canonical burnchain tip";
+               "height" => %canonical_burnchain_tip.block_height,
+               "block_hash" => %canonical_burnchain_tip.block_hash.to_string());
+
+        // Retrieve all the direct ancestors of this block with an unprocessed sortition
+        let (mut last_processed_ancestor, sortitions_to_process) =
+            self.find_sortitions_to_process(canonical_burnchain_tip.block_hash.clone())?;
+        let dbg_burn_header_hashes: Vec<_> = sortitions_to_process
+            .iter()
+            .map(|block| {
+                format!(
+                    "({}, {})",
+                    &block.header.block_hash.to_string(),
+                    block.header.block_height
+                )
+            })
+            .collect();
+
+        debug!(
+            "Unprocessed burn chain blocks [{}]",
+            dbg_burn_header_hashes.join(", ")
+        );
```

Comment:
It's probably more efficient debug print a string vector:

```suggestion
        debug!(
            "Unprocessed burn chain blocks {dbg_burn_header_hashes:?}"
        );
```